freestyle-tagless/shared/src/main/scala/tagless.scala

```diff
@@ -0,0 +1,146 @@
+
+      case Expr(cls: ClassDef) :: Expr(_) :: _ => fail(s"$invalid in ${cls.name}. $noCompanion")
+
+      case _ => fail(s"$invalid. $abstractOnly")
+    }
+
+
+    def mkEffectTrait(effectTrait: ClassDef): ClassDef = {
+      val FF = freshTypeName("FF$")
+      val requests: List[Request] = collectRequests(effectTrait)
+      val body = requests.map(_.traitDef(FF))
+      // this is to make a TypeDef for `$FF[_]`
+      val wildcard = TypeDef(Modifiers(Flag.PARAM), typeNames.WILDCARD, List(), TypeBoundsTree(EmptyTree, EmptyTree))
+      val ffTParam = TypeDef(Modifiers(Flag.PARAM), FF, List(wildcard), TypeBoundsTree(EmptyTree, EmptyTree))
+      val ClassDef(mods, name, tparams, Template(parents, self, _)) = effectTrait
+      ClassDef(mods, name, ffTParam :: tparams, Template(parents, self, body))
+    }
+
+    class Request(reqDef: DefDef) {
+
+      import reqDef.tparams
+
+      val reqImpl = TermName(reqDef.name.toTermName.encodedName.toString)
+
+      private[this] val Res = reqDef.tpt.asInstanceOf[AppliedTypeTree].args.last
+
+      val params: List[ValDef] = reqDef.vparamss.flatten
+
+      def freeDef: DefDef =
+        if (params.isEmpty)
+          q"def $reqImpl[..$tparams]: FS[$Res]"
+        else
+          q"def $reqImpl[..$tparams](..$params): FS[$Res]"
+
+      def freeHandlerDef(H: TermName, RT: TypeName): DefDef = {
+        val args = params.map(_.name)
+        if (params.isEmpty)
+          q"def $reqImpl[..$tparams]: $RT[$Res] = $H.${reqDef.name}(..$args)"
+        else
+          q"def $reqImpl[..$tparams](..$params): $RT[$Res] = $H.${reqDef.name}(..$args)"
+      }
+
+      def traitDef(FF: TypeName): DefDef =
+        if (params.isEmpty)
+          q"def $reqImpl[..$tparams]: $FF[$Res]"
+        else
+          q"def $reqImpl[..$tparams](..$params): $FF[$Res]"
+
+      def handlerDef: DefDef =
+        if (params.isEmpty)
+          q"def $reqImpl[..$tparams]: $MM[$Res]"
+        else
+          q"def $reqImpl[..$tparams](..$params): $MM[$Res]"
+
+    }
+
+    def collectRequests(effectTrait: ClassDef): List[Request] = effectTrait.impl.collect {
+      case dd@q"$mods def $name[..$tparams](...$paramss): $tyRes" => tyRes match {
+        case tq"FS[..$args]" => new Request(dd.asInstanceOf[DefDef])
+        case _ => fail(s"$invalid in definition of method $name in ${effectTrait.name}. $onlyReqs")
+      }
+    }
+
+    def mkEffectObject(effectTrait: ClassDef): ModuleDef = {
+
+      val requests: List[Request] = collectRequests(effectTrait)
+
+      val Eff = effectTrait.name
+      val TTs = effectTrait.tparams
+      val tns = TTs.map(_.name)
+      val ev = freshTermName("ev$")
+      val hh = freshTermName("hh$")
+      val stackSafeHandler = freshTermName("stackSafeHandler$")
+
+      q"""
+        object ${Eff.toTermName} {
+
+          trait Handler[$MM[_], ..$TTs] extends $Eff[$MM, ..$tns] {
+            ..${requests.map(_.handlerDef)}
+          }
+
+          @_root_.freestyle.free trait StackSafe {
```

Comment:
I wonder if, instead of having here the `@free` annotation, we could just extract the `StackSafe` part to another expression and make a call to the `freestyle.freeImpl.free`.